### DISCUSSION TITLE
Should I fill a bug report?

### DISCUSSION THREAD

**Discussion by thmxv** [bevyengine/bevy#14799] [Q&A] [answered]
```rust
use bevy::prelude::*;

fn test<T>(_items: &[T])
where
    for<'a> &'a T: IntoIterator<Item = &'a isize>,
{
}

pub fn system() {
    let items = vec![[1,2]];
    test(&items);
}
```

Causes and error: 
```
error[E0275]: overflow evaluating the requirement `&_: IntoIterator`
```
at line 11:
```rust
test(&items);
```

Removing the `use` statement make everything compile fine.

I found some related bug reports for the rust compiler but all are "resolved" (some last month) but I do get this error even if using the nightly toolchain.

Is the code OK? Should I fill a bug report? If yes, for bevy or for the rust compiler?

**Comment by MrGVSV** [comment]
Yeah a bug report would be nice. It might also help if you could narrow it down to a certain module or trait in the prelude that's causing this.

**Reply by thmxv** [reply]
Hi,

Thank you.

Should I report the issue for Bevy of for Rust?

Depending on the code to reproduce this issue (I modified it a lot to minimize the code), the type causing this in the prelude is: `Res` or `Local` or `Mut`. For the moment I have `Res`:
```
note: required for `&Res<'_, _>` to implement `IntoIterator`
```

**Reply by akimakinai** [reply, marked as answer]
The problem reproduces with the following code.

```rust
pub struct Local<'s, T>(&'s T);

impl<'s, 'a, T> IntoIterator for &'a Local<'s, T> where &'a T: IntoIterator {
    type Item = <&'a T as IntoIterator>::Item;
    type IntoIter = <&'a T as IntoIterator>::IntoIter;
    fn into_iter(self) -> Self::IntoIter {
        todo!()
    }
}

fn test<T>(_items: &[T])
where
    for<'a> &'a T: IntoIterator<Item = &'a isize>,
{
}

pub fn system() {
    let items = vec![[1,2]];
    test(&items);
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cac861a1305e1f5520ea1fadb3fd013c

This code compiles fine with `-Znext-solver`, so it may be a limitation of the current trait solver.

https://rust.godbolt.org/z/9fexfr4K9

**Reply by thmxv** [reply]
Great, thank you.

So this issue is not related to Bevy but the rust compiler and I suppose they are aware of it there too and I should not fill a bug report for the compiler neither.

Also thank you for the `-Znext-solver` workaround hopefully this is mature enough so that it will compile my project and dependencies.

**Reply by akimakinai** [reply]
Yes, it is a compiler issue. There are already [many issues](https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+overflow+evaluating+the+requirement) with "overflow evaluating the requirement" error, and some seem to be almost identical to this, e.g. https://github.com/rust-lang/rust/issues/106512, but I'm no expert.

**Reply by thmxv** [reply]
Ok, great. Thanks.

For those who wants to know: The `-Znext-solver` workaround does not work for a "real life" project:
- It required `nightly` which might be unstable
- The `-Znext-solver` option cannot be passed trough `cargo rustc` which would only use it for the base project but not for the dependencies
- Passing `-Znext-solver` to the compilation of all the dependencies does not work and cause errors while building some dependencies (in my case `bevy_utils`)

**Comment by thmxv** [comment]
I tried changing the "use" statement:
`use bevy;` causes the issue also
`use std::sync::Mutex;` or anything std does not causes the issue
`use a_non_bevy_module;` does not causes the issue

**Comment by ethereumdegen** [comment]
Is there at least some way to detect which line of code is throwing this error ?  I got this in a gigantic project and i have no way of figuring out the line of code responsible and im in a very very bad situation now

**Reply by thmxv** [reply]
For me the error messages contained the information of the line generating the error.

**Reply by cstegel** [reply]
I also get this error and would love to know what's causing it. Unfortunately, it came up while migrating my project from Bevy 0.14 -> 0.15 and making some necessary non-trivial changes. No line numbers or files are shown.

The `-Znext-solver=coherence` does not fix it for me:

```
RUSTFLAGS="-Znext-solver=coherence" rustup run nightly cargo build
```

```
$ rustup run nightly rustc --version
rustc 1.86.0-nightly (3f43b1a63 2025-01-03)
```

**Reply by cstegel** [reply]
I looked through the changes I needed to make while migrating from 0.14 -> 0.15 and found my cause: [AnimatableKeyFrameCurve::new](https://docs.rs/bevy/latest/bevy/animation/animation_curves/struct.AnimatableKeyframeCurve.html#method.new)

The arg to this function is `impl IntoIterator<Item = (f32, T)>`. I believe the compiler bug will be triggered for any file that has `use bevy::prelude::*` and calls a function that has an arg of type `impl IntoIterator`.

In my case, I still wanted to use `AnimatableKeyframeCurve::new` so my hack is to make a wrapper function in a separate file that doesn't have `use bevy::prelude::*`:

```rs
// new file, animatable_keyframe_curve.rs
use bevy::animation::animation_curves::AnimatableKeyframeCurve;
use bevy::prelude::cores::UnevenCoreError;
use bevy::prelude::Animatable;

pub fn new_animatable_keyframe_curve<T: Animatable>(
    keyframes: impl Iterator<Item = (f32, T)>,
) -> Result<AnimatableKeyframeCurve<T>, UnevenCoreError> {
    AnimatableKeyframeCurve::new(keyframes.into_iter())
}
```

```rs
// existing bevy code

use bevy::prelude::*;
use crate::animatable_keyframe_curve::new_animatable_keyframe_curve;
...
new_animatable_keyframe_curve([(0., Quat::IDENTITY), (1., Quat::IDENTITY)].into_iter());
```
